Step3 - 수강신청(DB적용)

namhyeop opened the pull request:
안녕하세요 한솔님!
Step3 수강신청 미션을 끝내서 PR을 올려요

제가 이번 미션 구현하면서 요구사항을 명확하게 이해하지 못한거 같아서 궁금한게 있는데
제가 해당 미션을 이해한거는 객체와 DB에 대한 매핑을 하는걸로 이해했는데
혹시 놓친 부분 있다면 피드백 부탁드려요!

ps.
추가적으로 Step2  MR 주시면서 피드백 주신부분도 포함해서 작업했습니다!

namhyeop committed:
feat : 시간 설정 로직 수정 (피드백 반영)

namhyeop committed:
feat : MethodSource 적용 (피드백 반영)

namhyeop committed:
feat : Session, Image 테이블 매핑

catsbi reviewed (changes requested):
남협님 3단계 미션은 실제로 해당 도메인들(ex: 강의, 과정등)이 database와 연결하고, 실제 테이블과 매핑하여 CRUD를 하는 API를 만들고 이를 테스트하는 과정입니다! 

관련한 기능들을 만들어서 다시 리뷰요청 부탁드려요!

namhyeop committed:
feat : DB 생성으로 인한 Sessions 삭제

namhyeop committed:
feat : DB 생성으로 인한 Sessions 삭제

namhyeop committed:
feat : JdbcSessionEntityRepository 생성

namhyeop committed:
feat : JdbcSessionEntityRepository 인터페이스 추가

namhyeop committed:
feat : JdbcSessionEntityRepository 테스트 코드 추가

namhyeop committed:
feat : NotExistSessionType 예외 추가

namhyeop committed:
feat : Session , 유효기간 객체 추가로 인한 테스트코드 추가 및 수정

namhyeop committed:
feat : 유효기간 객체 추가

namhyeop committed:
feat : ImageRepository 테스트 추가 및 구현

namhyeop committed:
feat : JdbcCourseRepository 패키지 경로 수정

namhyeop committed:
feat : JdbcCourseRepository 테스트 코드 추가

namhyeop committed:
feat : ImageEntity 추가

namhyeop committed:
feat : Image 객체 생성자 추가

namhyeop committed:
feat : FreeSession-생성자 추가

namhyeop committed:
feat : CourseRepositoryTest-패키지 경로 수정

namhyeop committed:
feat : CourseRepositoryTest-테스트 패키지 경로 수정

namhyeop committed:
feat : SessionFactory 구현

namhyeop committed:
feat : SessionFactory-SesstionType 구분을 위한 enum 추가

namhyeop committed:
feat : SessionEntity 추가

namhyeop committed:
feat : BaseEntity Getter 추가

namhyeop committed:
feat : Course 테스트코드 추가

namhyeop committed:
feat : Course registerSession 추가

namhyeop committed:
feat : Session, Image Mock 테스트코드 추가

namhyeop committed:
feat : SessionBuilder 추가

namhyeop committed:
feat : DB schema 추가

namhyeop committed:
feat : PaidSession 필드 수정으로 인한 테스트 코드 수정

namhyeop committed:
feat : PaidSession 생성자 추가

namhyeop committed:
feat : MaxRegistrationCount-Getter 추가

namhyeop committed:
feat : Money-Getter 추가

namhyeop committed:
feat : Session-freeSession을 위한 default 메소드 추가 및 DB 접근을 위한 Getter 추가

namhyeop committed:
feat : SessionRegisterFailException 추가

namhyeop committed:
feat : NotExistTimeException 추가

namhyeop committed:
feat : ValidityPeriodTest-NotExistTimeException 추가로 인한 테스트 코드 수정

namhyeop committed:
feat : Service 코드 추가

namhyeop committed:
feat : BaseEntity 패키지 경로 수정

namhyeop commented:
Session, Image 도메인 집중적으로 DB랑 매핑해서 작업을 진행해봤어요!
작업 진행하면서 JDBC Template으로 구현을 하다보니 시간이 좀 오래 걸렸네요

제가 이번 작업 진행하면서 domain 객체를 DB 매핑 + 도메인 기능 겸용으로 쓰는 부분 때문에 고민을 가장 많이 했어요. 
고민 때문에 "이걸 제대로 하는게 맞는건가?"라는 생각을 계속 하면서 구현했었는데요.. 

결과적으로는 도메인 객체를 DB 접근 용도와 같이 사용하는건 맞지 않다고 생각해서 Domain명 + "Entity"라는 이름의 DB 접근용 패키지를 하나 생성했습니다. (패키지명 : entity)

이렇게 나눈 이유가 도메인은 도메인 객체를 참조하면서 구성을 이루고 있는데 DB는 외래키를 가지고 DB를 참조 하다보니 Entity 패키지를 안 분리하니까 도메인은 도메인 객체를 참조하고 있음에도 ID 값도 가져야하는 구조가 나오더라고요. 그래서 코드 보실때 Entity 패키지는 DB 조회용 패키지이고 Domain 객체는 서비스 로직을 위해 존재하는 객체라고 봐주시면 될 거 같아요!

javajigi commented on src/main/java/nextstep/courses/domain/session/ValidityPeriod.java:
```diff
@@ -0,0 +1,33 @@
+import nextstep.courses.error.exception.EndDateBeforeStartDateException;
+import nextstep.courses.error.exception.NotExistTimeException;
+
+public class ValidityPeriod {
```
SessionPeriod라는 이름을 사용하는 것이 의미상 더 적절하지 않을까?

namhyeop replied:
안녕하세요 포비님!
SessionPeriod가 좀 더 자연스러운거 같네요!

근데 Session 객체 안에 들어가는거니까 Session은 빼고 Period는 어떻게 생각하세요?

javajigi commented on src/main/java/nextstep/courses/domain/session/impl/PaidSession.java:
```diff
@@ -1,18 +1,21 @@
 import nextstep.payments.domain.Money;
 import nextstep.payments.domain.Payment;
 
 public class PaidSession extends BaseEntity implements Session {
```
FreeSession과 PaidSession 사이에 많은 중복 코드가 존재한다.
중복을 제거하기 위해 FreeSession과 PaidSession와 Session 인터페이스 사이에 추상 클래스를 두는 것은 어떨까?
BaseEntity를 상속해야 한다면 앞에서 구현한 추상 클래스가 BaseEntity를 상속하면 되지 않을까?

namhyeop replied:
추상 클래스로 선언하는것이 적합한 설계인것 같아요

Session 구현할 때 이펙티브자바 아이템20(추상클래스보다는 인터페이스를 우선하라)을 참고해서 다중상속을 가능하도록 고려해서 인터페이스를 기준으로 구현하는 방향으로 설계했었어요.(추상 클래스는 다중 상속 불가능)

그런데 지금 구조는 Session이 다른 일급 컬렉션을 가지는 수직 구조의 설계인데 추상클래스를 사용안하려 하다보니 중복된 코드가 많아 진것 같아요.

그래서 피드백 주신것처럼 AbstractSession(추상클래스)을 하나 생성하고 AbstractSession이 Session 인터페이스를 상속받아서 다중 상속을 가능하다록 리팩토링 진행해볼게요(이런걸 골격 클래스라고 하네요)

javajigi commented on src/main/java/nextstep/courses/entity/SessionEntity.java:
```diff
@@ -0,0 +1,94 @@
+
+    public LocalDateTime getEndDate() {
+        return endDate;
+    }
```
이 Entity의 값을 활용해 Session을 생성하는 것은 어떨까?
그런데 Session만을 생성한다면 Session과 SessionEnttity에 중복 코드가 많아 보인다.
Session 객체를 더 작은 단위의 객체로 분리해 보는 것은 어떨까?
예를 들어 수강 신청을 담당하는 Enrollment 객체를 추가하고, SessionEntity가 Enrollment를 생성하도록 하는 것도 좋은 방법이 되지 않을까?

namhyeop replied:
Enrollment 객체를 추가해서 수강신청 로직 분리하도록 진행해보겠습니다

javajigi commented on src/test/java/nextstep/courses/domain/session/impl/FreeSessionTest.java:
```diff
@@ -65,7 +65,7 @@ class FreeSessionTest {
     void 무료_강의_수강신청시_강의_수강료가_0원이_아닐_경우_수강_신청이_불가능하다() {
```
수강 신청 로직을 테스트하기 위해 굳이 필요없는 이미지 정보 등과 같은 데이터까지 추가해야 한다.
수강 신청 로직을 담당하는 객체를 추가하고 이에 대한 단위 테스트를 진행해 보면 어떨까?
Session 객체의 크기가 너무 큰 것은 아닐까?

namhyeop replied:
이 부분도 Session에 대한 객체를 작게 쪼개는 작업 진행한뒤 단위테스트 추가해볼게요!

namhyeop replied:
이 부분 다시 생각해보니 수강신청할 때 Session의 상태를 체크하는 테스트인데 "무료_강의_수강신청시"라는 테스트명 때문에 혼선이 있었던거 같아서 테스트명을 수정해놓을게요!

javajigi commented on src/test/java/nextstep/courses/fixture/builder/PaidSessionBuilder.java:
```diff
@@ -0,0 +1,67 @@
+import nextstep.courses.domain.session.impl.FreeSession;
+import nextstep.payments.domain.Money;
+
+public class PaidSessionBuilder {
```
Builder 패턴 적용 👍

javajigi reviewed (changes requested):
도메인 객체와 DB 매핑 객체를 분리한 부분이 인상적이네요. 👍 
단, 도메인 객체(특히 Session)의 크기가 너무 크다보니 Entity와의 차이를 느끼기 어렵네요.
Entity는 현재와 같은 구조로 유지하고 도메인 객체를 더 작은 단위로 분리해 보면 어떨까요?

namhyeop committed:
feat : 인터페이스 -> 골격클래스로 리팩토링

namhyeop committed:
feat : ValidPeriod 객체명 수정

namhyeop committed:
feat : SessionFactory 패키지 변경

namhyeop committed:
feat : AbstractSession 위치 변경

namhyeop committed:
feat : 불필요한 import문 제거

namhyeop committed:
feat : Course 수강신청 로직 삭제

namhyeop committed:
feat : import패키지 경로 변경

namhyeop committed:
feat : Sessions 삭제로 인한 불필요한 equals, hascode 삭제

namhyeop committed:
feat : PaymentService @Service 추가

namhyeop committed:
feat : Enrollment 추가

namhyeop committed:
feat : FreeSessionTest 테스트명 수정

namhyeop committed:
feat : 불필요한 DTO 파일 삭제

namhyeop committed:
feat : SessionFactory 유료/무료 판단 로직 수정

namhyeop committed:
feat : Enrollment 객체 생성으로 인한 Course 테스트 로직 삭제

namhyeop committed:
feat : Enrollment 구현 및 테스트 코드 추가

namhyeop committed:
feat : Service 로직 수정

namhyeop committed:
feat : 불필요한 코드 삭제

javajigi commented on src/main/java/nextstep/courses/service/lmsService.java:
```diff
@@ -1,52 +1,45 @@
-
-        sessionRepository.updateRegistrationCount(SessionEntity.from(session, image));
-        return session;
+    public void enroll(Long sessionId, Enrollment enrollment){
+        SessionEntity sessionEntity = sessionRepository.findById(sessionId)
+            .orElseThrow(() -> new NotExistSession(sessionId));
+
+        Payment payment = paymentService.payment(enrollment);
+        Session session = SessionFactory.createSession(sessionEntity);
+
+        enrollment.enroll(session, payment);
+
+        sessionRepository.updateRegistrationCount(SessionEntity.from(session));
```
```suggestion
    public void enroll(Long sessionId, NsUser nsUser){
        SessionEntity sessionEntity = sessionRepository.findById(sessionId)
            .orElseThrow(() -> new NotExistSession(sessionId));
        Payment payment = paymentService.payment(sessionId, nsUser);
        Enrollment enrollment = sessionEntity.enrollment(); // 수강 신청 가능 여부 판단을 위한 모든 정보를 Enrollment에 저장
        Student student = enrollment.enroll(payment, nsUser);
        studentRepository.save(student); // Student 테이블에 새로운 수강생 정보를 추가하면 되지 않을까?
```

namhyeop replied:
포비님 코드로 피드백 주셔서 제가 잘못 이해한 부분 정확히 이해할 수 있었어요.
감사합니다.

저는 수강신청 등록에 대한 책임이 Session에 있어야 한다고 생각했는데 Enrollment로 옮기니 이전보다 수강 신청에 대한 테스트가 쉬워졌어요.

코드 수정했는데 이상한 부분있으면 다시 코멘트 주시면 감사할 것 같습니다!

javajigi reviewed (changes requested):
피드백 빠르게 반영했네요.
제가 피드백한 내용을 잘못 이해한 것 같아 해당 부분만 피드백 남겨봤어요.
제가 피드백한 내용은 수강신청만을 담당하는 도메인 객체를 추가해 보면 어떨까라는 제안이었습니다.

namhyeop committed:
feat : 불필요한 클래스 삭제

namhyeop committed:
feat : Enrollment 분리

namhyeop committed:
feat : 테스트 수정

namhyeop committed:
feat : DB 스키마 변경

namhyeop committed:
feat : 학생 데이터 추가

namhyeop committed:
feat : 구조 재설계로 인한 미사용 클래스 삭제

namhyeop committed:
feat : 구조 재설계 작업 반영

namhyeop committed:
feat : 테스트 실패 오류 수정

namhyeop committed:
feat : SessionFactory 메소드명 수정

namhyeop committed:
feat : sessionRepository 테스트 추가

namhyeop committed:
feat : sessionRepository 테스트 코드 추가

1. 비화설황된 sessionRepository 테스트 코드 활성화
2. 테스트를 위한 enrollmentBuilder 구현
3. session, sesssionEntiyu 생성자 추가

namhyeop committed:
feat : enrollment 테스트 추가 전 이전 테스트 코드 삭제

namhyeop committed:
feat : SessionRepositoryTest 클래스명 수정

namhyeop committed:
feat : Enrollment 테스트 코드 추가

namhyeop committed:
feat : EnrollmentCount 테스트 추가

namhyeop committed:
feat : codeConvention 적용 및 불 필요한 코드 제거

javajigi commented on src/main/java/nextstep/courses/domain/session/FreeSession.java:
```diff
@@ -0,0 +1,45 @@
+
+public class FreeSession extends Session {
+
+    public FreeSession(SessionEntity sessionEntity) {
```
Entity가 도메인 객체 의존 ok
도메인 객체 -> Entity에 의존하는 것은 괜찮나?

이렇게 의존 관계를 맺을 경우 entity 패키지와 domain 패키지가 cyclic dependency가 생긴다.
도메인 객체가 Entity 객체에 의존하는 것에 대한 한번 고민해 보면 어떨까?

namhyeop replied:
Cyclic Dependency 관련해서 피드백 주셔서 감사합니다!
교육으로만 학습하고 적용해볼 생각을 못했네요.

기존에 사용중인 Factory에 Session 관련 생성 로직을 추가해서 수정해봤습니다.

피드백 적용하면서 질문이 한 가지 있습니다.
Cyclic Dependency가 발생했다라는 기준을 알 수 있을까요?

![스크린샷 2024-04-26 오후 10 46 54](https://github.com/next-step/java-lms/assets/53933920/a047e1bb-a1f1-4dcb-94d3-a379c51107ac)
강의에서 위 그림에서 Cyclic Dependency가 발생한다고 하셨는데 이유가 이해가지 않습니다.
저는 의존성의 기준이 import문으로 알고 있는데 사진을 보시면 NextStepLine은  Line 인터페이스를 정의하는데 정의를 한다고 해서 import문이 생기지는 않기 때문에 Cyclic Dependency가 생기는 것이 아닌가라고 생각 하고 있습니다.

아니면 Cyclic Dependency의 기준은 import문이 아니라 해당 클래스를 사용하는 여부일까요?

javajigi replied:
import문을 사용하는 것 맞습니다.
위 그림에서는 드러나지 않는데요.
인터페이스 기반으로 구현하다보면 인터페이스와 구현 클래스를 사용해 최종적으로 사다리를 완성할 때 Cyclic Dependency가 발생하는 지점이 있다는 것을 표현하려고 했어요.
저 그림만 봐서는 해당 부분이 보이지 않는 건 맞습니다.

namhyeop replied:
답변 감사드립니다!
궁금한게 해결되었네요

javajigi commented on src/main/java/nextstep/courses/domain/session/enrollment/Enrollment.java:
```diff
@@ -0,0 +1,28 @@
+import nextstep.payments.domain.Payment;
+import nextstep.users.domain.NsUser;
+
+public interface Enrollment {
```
도메인 객체를 인터페이스로 구현하고 있다.
그런데 굳이 Enrollment와 같은 객체도 인터페이스로 구현할 필요가 있을까?
굳이 불필요한 곳까지 남용하는 것은 아닐까?

namhyeop replied:
Enrollment는 등록 관련된 책임과 정보를 담고있는 객체여서 추후에 누군가 정책이 변경되어 재정의를 하게 된다면 아래 메소드를 재정의하길 바라는 바람에 인터페이스를 정의했는데 이런 경우는 어떻게 생각하세요?

1. 수강 신청
2. 수강 신청 가능한 조건 정책인지
3. 강의 상태가 수강신청이 가능한지
4. 결제 조건이 만족하는지

javajigi replied:
너무 섣부른 추상화가 아닐까요?
서비스를 유지보수하면서 정책 변경이 발생할 때 빠르게 리팩터링하는 것도 좋지 않을까 생각해요.
미래에 어떤 부분이 어떻게 정책이 변경될지 미리 파악하기 너무 어렵잖아요.

namhyeop replied:
혹시 추상화를 하시는 기준이 있으실까요?

javajigi reviewed (changes requested):
피드백 반영 잘 했네요. 💯 
바로 merge하려다 중요하다 생각하는 부분이 두 가지 있어 피드백 남겨봤어요.
피드백 남긴 부분에 대해 한번 고민해 보면 어떨까요?

namhyeop committed:
feat : Image <-> ImageEntity 의존성 제거

namhyeop committed:
feat : Session <-> SessionEntity Cyclic Dependency제거

javajigi reviewed (approved):
피드백 잘 반영했네요. 💯 
질문에 대해서는 해당 질문에 답글로 남겼어요.
이번 3단계가 쉽지 않은 과정이었을텐데 빠르게 피드백 반영하면서 잘 마무리했네요.
4단계는 가능한 안전한 방법(스트랭귤러 패턴 적용)으로 리팩터링 도전해 보면 어떨까요?
